library_information_validator.py

```diff
@@ -0,0 +1,495 @@
+        error_missing_index = 0
+        warning_mixed_indexes = 0
+        sindex_chosen = True
+        if(len(cindex_absent) == len(pool_rows)):
+            sel_index = sindex_list
+            if(len(sindex_list) != len(pool_rows)):
+                for absent_index in sindex_absent:
+                    logger.error("missing index in row {}".format(re.sub("\D","",absent_index)))
+                    quit()
+        elif(len(sindex_absent) == len(pool_rows)):
+            sel_index = cindex_list
+            sindex_chosen = False
+            if(len(cindex_list) != len(pool_rows)):
+                for absent_index in cindex_absent:
+                    logger.error("missing index in row {}".format(re.sub("\D","",absent_index)))
+                    quit()
+        else:
+            sindex_chosen = False
+
+            # warning if a pool consists out of a mix of NGI standard indexes and custom indexes
+            sel_index = sindex_list + cindex_list
+            if(len(sel_index) != len(pool_rows)):
+                rowID_cindex_absent = [re.sub("\D","",x) for x in cindex_absent]
+                rowID_sindex_absent = [re.sub("\D","",x) for x in sindex_absent]
+
+                missing_indexes = list(set(rowID_cindex_absent).intersection(rowID_sindex_absent))
+                for mIndex in missing_indexes:
+                    logger.error('missing index in row {}'.format(mIndex))
+
+            logger.warning('mix between custom and standard indexes in pool {}.'\
+            .format(pool))
+            warning_mixed_indexes += 1
+
+        # returns warnings for check summary
+        return(sel_index, sindex_chosen, warning_mixed_indexes, warning_component_length)
+
+    def validate_index(self, index_seq, pool_name, sindex):
+        '''
+        does all the fancy index checks
+        '''
+
+        # allows for entry "noIndex" if only one sample is defined in the pool
+        c = Counter(index_seq)
+        if(c['noIndex'] > 0 and len(index_seq) != 1):
+            logger.error('Pool {} contains undefined index(es) (\"noIndex\")'\
+            .format(pool_name))
+            quit()
+        elif(c['noIndex'] == 1 and len(index_seq) == 1):
+            logger.info('Pool {} containing one sample is not indexed.'\
+            .format(pool_name))
+        elif(c['noIndex'] == 0):
+            # checks that all indexes in a pool are unique
+            for index, index_count in c.most_common():
+                if(index_count>1):
+                    logger.error('The index sequence \"{}\" in pool {} is not unique for this pool.'\
+                    .format(index, pool_name))
+                    quit()
+                else:
+                    break
+
+        warning_low_div = 0
+        index_count = 1
+        for index in index_seq:
+            # checks that indexes only contain valid letters
+            charRE = re.compile(r'[^ATCGNatcgn\-.]')
```

Comment:
The regex doesn't change between iterations so you should probably compile it outside the loop.

Reply:
I was about to point out that `.` is a special regex character which means "any character" and that your regex was broken. However it seems that if it's within `[^.]` then it's a special case where it's _not_ a special character. Anyway, I would escape it to help future people reading your code 😅 (it should still work the same when escaped)

```python
charRE = re.compile(r'[^ATCGNatcgn\-\.]')
```